tests/common/user_test.go

```diff
@@ -0,0 +1,371 @@
+// Copyright 2022 The etcd Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package common
+
+import (
+	"testing"
+	"time"
+
+	"github.com/stretchr/testify/assert"
+	"go.etcd.io/etcd/tests/v3/framework/config"
+	"go.etcd.io/etcd/tests/v3/framework/testutils"
+)
+
+func TestUserAdd_Simple(t *testing.T) {
+	testRunner.BeforeTest(t)
+	tcs := []struct {
+		name   string
+		config config.ClusterConfig
+	}{
+		{
+			name:   "NoTLS",
+			config: config.ClusterConfig{ClusterSize: 1},
+		},
+		{
+			name:   "PeerTLS",
+			config: config.ClusterConfig{ClusterSize: 3, PeerTLS: config.ManualTLS},
+		},
+		{
+			name:   "PeerAutoTLS",
+			config: config.ClusterConfig{ClusterSize: 3, PeerTLS: config.AutoTLS},
+		},
+		{
+			name:   "ClientTLS",
+			config: config.ClusterConfig{ClusterSize: 1, ClientTLS: config.ManualTLS},
+		},
+		{
+			name:   "ClientAutoTLS",
+			config: config.ClusterConfig{ClusterSize: 1, ClientTLS: config.AutoTLS},
+		},
+	}
+	for _, tc := range tcs {
+		nestedCases := []struct {
+			name          string
+			username      string
+			password      string
+			noPassword    bool
+			expectError   bool
+			expectedError string
+		}{
+			{
+				name:        "empty_username_not_allowed",
+				username:    "",
+				password:    "foobar",
+				expectError: true,
+				// Very Vague error expectation because the CLI and the API return very
+				// different error structures.
+				expectedError: "user name",
```

Comment:
Please consider to combine the two fields "expectEror" and "expectedError" into one with type error. If it's nil, then it means we expect no error.

Reply:
I think the problem that we cannot compare errors in e2e tests as we just get stdout. It's worth improving, however out of scope of this PR.

Reply:
This is a common test, it shouldn't care about whether it's running in e2e or integration environment. It just calls an interface method (`UserAdd` in this example, see below), and we should can always depend on the second return  value to check the success of the operation.

`UserAdd(name, password string, opts config.UserAddOptions) (*clientv3.AuthUserAddResponse, error)`

Reply:
The existence of an interface doesn't mean that all implementations will return the same error in this case - because the CLI and API are two very different entities here today.

Because of this, we require some acknowledgement of them being different (in this case, that involves doing string comparisons on errors, as much as it would be nicer not to).

Reply:
```suggestion
// remove expectError, and change expectedError's type to error

if expectedError != nil {
      // it means we expected some error
      // compare the returned error message and expected error message
} else {
      // We expect no error
}				
```